File: frontend/src/types/video.ts
import { ApiResponse } from './api';
import { Action, Base, Rateable } from './common';
import { PublicUser } from './user';

export enum UploadState {
  Uploading = 'uploading',
  Done = 'done',
  Failed = 'failed',
  Processing = 'processing',
}

export interface Video extends Base {
  title: string;
  description: string;
  public: boolean;
  tags: string;
  filename: string;

  owner: PublicUser;

  duration: number;
  minViewDuration: number;

  views: number;

  dislikesCount: number;
  likesCount: number;

  isLiked: boolean;
  isDisliked: boolean;
}

export interface VideoAction extends Action {
  videoId: number;
  user: PublicUser;
}

export interface UploadBody {
  filename: string;
  title: string;
  size: number;
  duration: number;
}

export interface UploadChunkBody {
  from: number;
  to: number;
  bytes: string;
  taskId: number;
}

export interface UploadChunkResponse {
  next?: UploadNextChunk;
  data: UploadTask;
  msg?: string;
}

export type UploadProgressResponse = ApiResponse<UploadTask>;

export interface UploadTask extends VideoAction {
  status: UploadState;
  uploaded: number;
  size: number;
  filename: string;
}

export interface UploadNextChunk {
  from: number;
  to: number;
  taskId: number;
}

export interface VideoComment extends VideoAction {
  text: string;

  isLiked: boolean;
  isDisliked: boolean;
  isHearted: boolean;
  likeCount: number;
  dislikeCount: number;
  replyCount: number;
}

export interface Reply extends Rateable, Base {
  text: string;
  videoId: number;
  commentId: number;
  userId: number;
  user: PublicUser;
}

export type CreateVideoCommentBody = Pick<VideoComment, 'text'>;
export type CreateVideoCommentResponse = ApiResponse<VideoComment>;

export type UpdateVideoCommentBody = Pick<VideoComment, 'text'>;
export type UpdateVideoCommentResponse = ApiResponse<VideoComment>;

export type GetVideoCommentResponse = ApiResponse<
  Array<VideoComment>,
  { totalCount: number; pinned?: number }
>;

export interface Playlist extends Base {
  title: string;
  description: string;
  public: boolean;
  ownerId: number;
  owner: PublicUser;
}

export interface PlaylistVideo extends Base {
  videoId: number;
  video: Video;
  playlistId: number;
  index: number;
}

File: backend/schema/video.go
package schema

// ! camelCase json fields !
type Video struct {
	Title           string  `json:"title" gorm:"not null"`
	Description     string  `json:"description" gorm:"not null"`
	Public          bool    `json:"isPublic" gorm:"default:false"`
	OwnerId         int     `json:"ownerId" gorm:"not null"`
	Owner           User    `json:"owner" gorm:"foreignKey:OwnerId"`
	Tags            string  `json:"tags"`
	Filename        string  `json:"filename"`
	Duration        float64 `json:"duration" gorm:"default:0"`
	MinViewDuration float64 `json:"minViewDuration" gorm:"default:30"`
	LikesCount      int     `json:"likesCount" gorm:"-"`
	DisLikesCount   int     `json:"dislikesCount" gorm:"-"`
	IsLiked         bool    `json:"isLiked" gorm:"-"`
	IsDisLiked      bool    `json:"isDisliked" gorm:"-"`
	Views           int64   `json:"views" gorm:"default:0"`

	// gorm overrides
	Base
}

type VideoAction struct {
	VideoId int   `json:"videoId" gorm:"not null"`
	Video   Video `json:"video" gorm:"foreignKey:VideoId"`

	Action
}

type AnonymousVideoAction struct {
	Base

	UserId  int   `json:"userId"`
	VideoId int   `json:"videoId" gorm:"not null"`
	Video   Video `json:"video" gorm:"foreignKey:VideoId"`
}

type VideoUploadTask struct {
	Status   string `json:"status" gorm:"not null"`
	Uploaded int64  `json:"uploaded" gorm:"default:0"`
	Size     int64  `json:"size" gorm:"not null"`
	Filename string `json:"filename" form:"not null"`

	VideoAction
}

type VideoUploadChunk struct {
	From     int64  `json:"-" gorm:"not null"`
	To       int64  `json:"-" gorm:"not null"`
	Filename string `json:"-" gorm:"no null"`

	VideoAction
}

type VideoWatchTime struct {
	Time float64 `json:"time" gorm:"not null"`
	From float64 `json:"from" gorm:"not null"`
	To   float64 `json:"to" gorm:"not null"`

	AnonymousVideoAction
}

type VideoView struct {
	AnonymousVideoAction
}

type VideoLike struct {
	VideoAction
}

type VideoDisLike struct {
	VideoAction
}

type VideoComment struct {
	Text string `json:"text" gorm:"not null"`

	ReplyCount int64 `json:"replyCount"  gorm:"default:0"`

	Rateable
	VideoAction
}

type VideoCommentAction struct {
	Comment   VideoComment `json:"comment" gorm:"foreignKey:CommentId"`
	CommentId int          `json:"commentId" gorm:"not null"`

	VideoId int   `json:"videoId" gorm:"not null"`
	Video   Video `json:"video" gorm:"foreignKey:VideoId"`

	Action
}

type VideoPinnedComment struct {
	VideoCommentAction
}

type VideoCommentLike struct {
	VideoCommentAction
}

type VideoCommentDisLike struct {
	VideoCommentAction
}

type VideoCommentReply struct {
	Text string `json:"text" gorm:"not null"`

	Rateable
	VideoCommentAction
}

type VideoCommentReplyAction struct {
	Reply   VideoCommentReply `json:"reply" gorm:"foreignKey:ReplyId"`
	ReplyId int               `json:"replyId" gorm:"not null"`

	VideoCommentAction
}

type VideoCommentReplyLike struct {
	VideoCommentReplyAction
}

type VideoCommentReplyDisLike struct {
	VideoCommentReplyAction
}
